components/tidb_query/src/expr_util/collation/mod.rs

```diff
@@ -55,7 +55,22 @@ impl From<DecodeOrWriteError> for crate::error::EvaluateError {
     }
 }
 
+pub trait Charset {
+    fn advance_one(data: &[u8]) -> usize;
```

Comment:
The current implementation of `advance_one` looks like `next_char_width` to me... while an `advance_one` method should be something like `fn advance_one(data: &mut [u8]) -> usize` which should really advance the input slice and return the char codepoint.

I'm not sure what an optimal interface would be here.